namer/namer.cc

```diff
@@ -2324,7 +2324,7 @@ vector<ast::ParsedFile> symbolizeTrees(const core::GlobalState &gs, vector<ast::
         for (auto result = fileq->try_pop(job); !result.done(); result = fileq->try_pop(job)) {
             if (result.gotItem()) {
                 Timer timeit(gs.tracer(), "naming.symbolizeTreesOne", {{"file", string(job.file.data(gs).path())}});
-                core::Context ctx(gs, core::Symbols::root(), job.file);
+                core::MutableContext ctx(gs, core::Symbols::root(), job.file);
```

Comment:
This change (and accompanying changes to `TreeSymbolizer`) are not OK.  I understand that the mutation of `GlobalState` that's necessary here is confined to a single bitflag, and that racy writes to the bitflag might theoretically be OK given hardware and compiler behavior.  However:

1. There's no guarantee that future modifications to `TreeSymbolizer` will be so careful; we generally rely on the types of `*Context` (and `const GlobalState` vs. `GlobalState`) to provide (some) safety guarantees when writing parallel code.  It would be very easy to come along and use some non-const method on `ctx` or `ctx.state`, not realizing that _this_ `MutableContext` is actually being used in a threaded context, and that would lead to some very difficult to diagnose bugs.
2. The racy modification to `isBehaviorDefining` is undefined behavior according to the C++ spec, so even if it doesn't cause problems today, it could tomorrow (and/or make running with ThreadSanitizer even harder than it is today).

I would also have to think about whether this flag still works correctly in the presence of the LSP fast path; I think it does today, because class changes take the slow path, but this flag would be a trap for the unwary if we ever made class changes take the fast path.

Reply:
Yep, I agree, I was going to back this out and come up with another way. Can discuss today.

Reply:
And yeah, it makes some LSP fast path tests fail